File: apps/server-admin/src/lib/system-notifications/providers/bark.ts
import type {
  NotificationDispatchContext,
  NotificationMessage,
  NotificationProvider,
  NotificationProviderDefinition,
  NotificationSchemaField,
  NotificationSendResult,
} from "../types";
import {
  resolveOptionalNonNegativeInteger,
  resolvePrimaryActionUrl,
  splitCommaSeparatedValues,
  toPlainRecord,
  toTrimmedString,
  truncateText,
} from "./shared";
import { tDefault } from "../../i18n";

const barkT = (
  key: string,
  params?: Record<string, string | number | boolean | null | undefined>,
) => tDefault(`server.notifications.providers.catalog.bark.${key}`, params);

const BARK_CONNECTION_SCHEMA: NotificationSchemaField[] = [
  {
    key: "server_url",
    label: barkT("fields.server_url.label"),
    description: barkT("fields.server_url.description"),
    placeholder: "https://api.day.app",
    type: "string",
    required: true,
    default_value: "https://api.day.app",
  },
  {
    key: "device_key",
    label: "Device Key",
    description: barkT("fields.device_key.description"),
    placeholder: "ynJ5Ft4atkMkWeo2PAvFhF",
    type: "string",
    required: true,
    sensitive: true,
  },
  {
    key: "timeout_seconds",
    label: barkT("fields.timeout_seconds.label"),
    type: "number",
    required: true,
    default_value: 5,
    min: 1,
    max: 30,
  },
];

const BARK_TARGET_SCHEMA: NotificationSchemaField[] = [
  {
    key: "level",
    label: barkT("fields.level.label"),
    description: barkT("fields.level.description"),
    type: "select",
    default_value: "active",
    options: [
      { label: "active", value: "active" },
      { label: "timeSensitive", value: "timeSensitive" },
      { label: "passive", value: "passive" },
      { label: "critical", value: "critical" },
    ],
  },
  {
    key: "group",
    label: barkT("fields.group.label"),
    description: barkT("fields.group.description"),
    placeholder: "fn-knock",
    type: "string",
  },
  {
    key: "sound",
    label: barkT("fields.sound.label"),
    description: barkT("fields.sound.description"),
    placeholder: "alarm",
    type: "string",
  },
  {
    key: "url",
    label: barkT("fields.url.label"),
    description: barkT("fields.url.description"),
    placeholder: "https://example.com/events/123",
    type: "string",
  },
  {
    key: "icon",
    label: barkT("fields.icon.label"),
    description: barkT("fields.icon.description"),
    placeholder: "https://day.app/assets/images/avatar.jpg",
    type: "string",
  },
  {
    key: "badge",
    label: barkT("fields.badge.label"),
    description: barkT("fields.badge.description"),
    type: "number",
    min: 0,
    max: 99999,
  },
  {
    key: "call",
    label: barkT("fields.call.label"),
    description: barkT("fields.call.description"),
    type: "boolean",
    default_value: false,
  },
];

export const barkProviderDefinition: NotificationProviderDefinition = {
  type: "bark",
  label: "Bark",
  description: barkT("description"),
  connection_schema: BARK_CONNECTION_SCHEMA,
  target_schema: BARK_TARGET_SCHEMA,
  sensitive_fields: ["device_key"],
  capabilities: {
    supports_text: true,
    supports_markdown: false,
    supports_rich_blocks: false,
    supports_actions: true,
    supports_mentions: false,
    supports_attachments: false,
    supports_provider_dedupe_key: false,
    max_body_length: null,
  },
};

const resolveBarkUrl = (provider: NotificationProvider) => {
  const baseUrl = toTrimmedString(provider.connection_config.server_url);
  const normalizedBaseUrl = baseUrl || "https://api.day.app";
  return `${normalizedBaseUrl.replace(/\/+$/, "")}/push`;
};

const buildBarkPayload = (
  message: NotificationMessage,
  context?: Partial<NotificationDispatchContext>,
) => {
  const targetConfig = toPlainRecord(context?.target?.target_config);
  const summary = toTrimmedString(message.summary);
  const bodyText = toTrimmedString(message.body_text);
  const hasStandaloneBody = Boolean(bodyText) && bodyText !== summary;
  const title = toTrimmedString(message.title || barkT("message.fallbackTitle"));
  const subtitle = hasStandaloneBody ? summary : "";
  const body = hasStandaloneBody ? bodyText : summary || bodyText || title;
  const url =
    toTrimmedString(targetConfig.url) || resolvePrimaryActionUrl(message);
  const level = toTrimmedString(targetConfig.level || "active");
  const sound = toTrimmedString(targetConfig.sound);
  const group = toTrimmedString(targetConfig.group);
  const icon = toTrimmedString(targetConfig.icon);
  const badge = resolveOptionalNonNegativeInteger(targetConfig.badge);
  const call = Boolean(targetConfig.call);

  return {
    title,
    subtitle: subtitle || undefined,
    body: body || barkT("message.fallbackTitle"),
    level: level || "active",
    ...(sound ? { sound } : {}),
    ...(group ? { group } : {}),
    ...(url ? { url } : {}),
    ...(icon ? { icon } : {}),
    ...(badge !== undefined ? { badge } : {}),
    ...(call ? { call: "1" } : {}),
  };
};

const sendSingleBarkPush = async (args: {
  url: string;
  deviceKey: string;
  message: NotificationMessage;
  context?: Partial<NotificationDispatchContext>;
  timeoutSeconds: number;
}) => {
  const payload = {
    ...buildBarkPayload(args.message, args.context),
    device_key: args.deviceKey,
  };
  const controller = new AbortController();
  const timeout = setTimeout(
    () => controller.abort(),
    Math.max(1, args.timeoutSeconds) * 1000,
  );

  try {
    const response = await fetch(args.url, {
      method: "POST",
      headers: {
        "content-type": "application/json; charset=utf-8",
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });
    const responseText = await response.text().catch(() => "");
    let parsedResponse: {
      code?: number;
      message?: string;
      timestamp?: number;
    } | null = null;
    try {
      parsedResponse = responseText ? JSON.parse(responseText) : null;
    } catch {
      parsedResponse = null;
    }

    const barkCode = parsedResponse?.code;
    const succeeded =
      response.ok && (barkCode === undefined || barkCode === 200);
    const reason =
      parsedResponse?.message ||
      (response.ok ? "" : `Bark returned ${response.status}`);

    return {
      success: succeeded,
      retryable:
        !succeeded && (response.status >= 500 || response.status === 429),
      reason: succeeded ? undefined : reason,
      response_summary: {
        status: response.status,
        ok: response.ok,
        code: barkCode,
        message: parsedResponse?.message,
        body_preview: truncateText(responseText),
      },
    };
  } catch (error) {
    return {
      success: false,
      retryable: true,
      reason:
        error instanceof Error ? error.message : barkT("errors.requestFailed"),
      response_summary: null,
    };
  } finally {
    clearTimeout(timeout);
  }
};

export const sendBarkMessage = async (args: {
  provider: NotificationProvider;
  message: NotificationMessage;
  context?: Partial<NotificationDispatchContext>;
  timeoutSeconds: number;
}): Promise<NotificationSendResult> => {
  const url = resolveBarkUrl(args.provider);
  const deviceKeys = splitCommaSeparatedValues(
    args.provider.connection_config.device_key,
  );
  if (!deviceKeys.length) {
    return {
      success: false,
      retryable: false,
      reason: barkT("errors.missingDeviceKey"),
    };
  }

  const payloadPreview = buildBarkPayload(args.message, args.context);
  const results = await Promise.all(
    deviceKeys.map((deviceKey) =>
      sendSingleBarkPush({
        url,
        deviceKey,
        message: args.message,
        context: args.context,
        timeoutSeconds: args.timeoutSeconds,
      }),
    ),
  );

  const failedResults = results.filter((result) => !result.success);
  if (!failedResults.length) {
    return {
      success: true,
      retryable: false,
      request_summary: {
        method: "POST",
        url,
        device_key_count: deviceKeys.length,
        level: payloadPreview.level,
        group: payloadPreview.group,
        title_preview: payloadPreview.title,
      },
      response_summary: {
        success_count: results.length,
        failed_count: 0,
        results: results.map((result) => result.response_summary),
      },
    };
  }

  return {
    success: false,
    retryable: failedResults.some((result) => result.retryable),
    reason:
      failedResults.length === 1
        ? failedResults[0]!.reason || barkT("errors.pushFailed")
        : barkT("errors.targetsFailed", {
            failed: failedResults.length,
            total: results.length,
          }),
    request_summary: {
      method: "POST",
      url,
      device_key_count: deviceKeys.length,
      level: payloadPreview.level,
      group: payloadPreview.group,
      title_preview: payloadPreview.title,
    },
    response_summary: {
      success_count: results.length - failedResults.length,
      failed_count: failedResults.length,
      results: results.map((result) => ({
        success: result.success,
        retryable: result.retryable,
        reason: result.reason,
        response_summary: result.response_summary,
      })),
    },
  };
};

File: apps/server-admin-rs/src/notifications/routes/providers/bark.rs
use super::*;

pub(in crate::notifications::routes) fn bark_definition() -> ProviderDefinition {
    ProviderDefinition {
        provider_type: "bark",
        label: "Bark",
        description: "Send notifications through Bark.",
        connection_schema: vec![
            string_schema(
                "server_url",
                "Server URL",
                true,
                false,
                Some("https://api.day.app"),
            )
            .placeholder("https://api.day.app"),
            string_schema("device_key", "Device Key", true, true, None)
                .placeholder("ynJ5Ft4atkMkWeo2PAvFhF"),
            number_schema("timeout_seconds", "Timeout seconds", true, Some(5)).bounds(1, 30),
        ],
        target_schema: vec![
            select_schema(
                "level",
                "Level",
                false,
                Some("active"),
                &["active", "timeSensitive", "passive", "critical"],
            ),
            string_schema("group", "Group", false, false, None).placeholder("fn-knock"),
            string_schema("sound", "Sound", false, false, None).placeholder("alarm"),
            string_schema("url", "URL", false, false, None)
                .placeholder("https://example.com/events/123"),
            string_schema("icon", "Icon", false, false, None)
                .placeholder("https://day.app/assets/images/avatar.jpg"),
            number_schema("badge", "Badge", false, None).bounds(0, 99999),
            bool_schema("call", "Call", false, Some(false)),
        ],
        sensitive_fields: vec!["device_key"],
        supports_markdown: false,
        supports_actions: true,
        supports_mentions: false,
        supports_provider_dedupe_key: false,
    }
}

pub(in crate::notifications::routes) async fn send_bark(
    state: &AppState,
    provider: &Value,
    target: &Value,
    message: &Value,
    timeout_seconds: i64,
) -> ProviderTestResult {
    let config = provider_config(provider);
    let device_keys = split_values(config.get("device_key"));
    if device_keys.is_empty() {
        return missing_config_result(&notification_provider_error_default(
            "bark",
            "missingDeviceKey",
            &[],
        ));
    }
    let base_url = default_string(config_text(&config, "server_url"), "https://api.day.app");
    let url = format!("{}/push", base_url.trim_end_matches('/'));
    let payload_preview = build_bark_payload(message, target);
    let mut results = Vec::new();
    for device_key in &device_keys {
        let mut payload = payload_preview.clone();
        insert_string(&mut payload, "device_key", device_key.clone());
        let (status, ok, text, parsed) = post_json(state, &url, &payload, timeout_seconds).await;
        let bark_code = parsed.as_ref().and_then(|value| json_i64(value, "code"));
        let success = ok && bark_code.is_none_or(|code| code == 200);
        let retryable = !success && (status >= 500 || status == 429);
        let reason = parsed
            .as_ref()
            .and_then(|value| json_text(value, "message"))
            .unwrap_or_else(|| {
                if status == 599 && !text.is_empty() {
                    return text.clone();
                }
                if ok {
                    String::new()
                } else {
                    format!("Bark returned {status}")
                }
            });
        let response_summary = if status == 599 && parsed.is_none() {
            Value::Null
        } else {
            json!({
            "status": status,
            "ok": ok,
            "code": bark_code,
            "message": parsed.as_ref().and_then(|value| json_text(value, "message")),
            "body_preview": truncate_text(&text, 500)
            })
        };
        let mut result = Map::new();
        result.insert("success".to_string(), Value::Bool(success));
        result.insert("retryable".to_string(), Value::Bool(retryable));
        if !success && !reason.is_empty() {
            result.insert("reason".to_string(), Value::String(reason));
        }
        result.insert("response_summary".to_string(), response_summary);
        results.push(Value::Object(result));
    }
    let failed_count = results
        .iter()
        .filter(|result| result.get("success").and_then(Value::as_bool) != Some(true))
        .count();
    let response_results = if failed_count == 0 {
        results
            .iter()
            .map(|result| {
                result
                    .get("response_summary")
                    .cloned()
                    .unwrap_or(Value::Null)
            })
            .collect::<Vec<_>>()
    } else {
        results.clone()
    };
    ProviderTestResult {
        success: failed_count == 0,
        retryable: results.iter().any(|result| {
            result
                .get("retryable")
                .and_then(Value::as_bool)
                .unwrap_or(false)
        }),
        message: if failed_count == 0 {
            notification_service_default_text("testSendSuccess", &[])
        } else if failed_count == 1 {
            results
                .iter()
                .find(|result| result.get("success").and_then(Value::as_bool) != Some(true))
                .and_then(|result| result.get("reason"))
                .and_then(Value::as_str)
                .filter(|reason| !reason.is_empty())
                .map(str::to_string)
                .unwrap_or_else(|| notification_provider_error_default("bark", "pushFailed", &[]))
        } else {
            format!("Bark failed for {failed_count}/{} target(s)", results.len())
        },
        request_summary: Some(json!({
            "method": "POST",
            "url": url,
            "device_key_count": device_keys.len(),
            "level": payload_preview.get("level").cloned().unwrap_or(Value::Null),
            "group": payload_preview.get("group").cloned().unwrap_or(Value::Null),
            "title_preview": payload_preview.get("title").cloned().unwrap_or(Value::Null)
        })),
        response_summary: Some(json!({
            "success_count": results.len().saturating_sub(failed_count),
            "failed_count": failed_count,
            "results": response_results
        })),
    }
}
